<a href="https://colab.research.google.com/github/VenodiW/Machine-Learning/blob/main/Convolutional_Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Convolutional neural network - Fashion-MNIST

## Table of Content
   
1. [Data Description](#cell_Preparation)

2. [Data Preprocessing](#cell_Preprocessing)

3. [Model Construction](#cell_model)

4. [Model Execution](#cell_execution)
 
5. [Experiments Report](#cell_report)

## Executive Summary

*The 'fashion-mnist_train' dataset has been used to train the neural network model to classify the clothing images *

*According to the model explained below, the sum of errors made for each example in training or validation sets is close to 4.1323 and the model has a 75.58% of accuracy with the predicted data compared to the true data.*

<a id = "cell_Preparation"></a>
## 1. Data Description

Fashion-MNIST is a dataset of Zalando's article images—consisting of a training set of 60,000 examples. Each example is a 28x28 grayscale image,
associated with a label from 10 classes. The dataset is consisting of 785 pixel coloums.

In [ ]:
#Importing the libraries
import keras
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation, Flatten
from keras.layers.convolutional import Convolution2D, MaxPooling2D
from keras.utils import np_utils
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np 

In [ ]:
#Importing the dataset
fashion = pd.read_csv('fashion-mnist_train.csv')
print(fashion.head())
fashion.shape

   label  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  \
0      2       0       0       0       0       0       0       0       0   
1      9       0       0       0       0       0       0       0       0   
2      6       0       0       0       0       0       0       0       5   
3      0       0       0       0       1       2       0       0       0   
4      3       0       0       0       0       0       0       0       0   

   pixel9  ...  pixel775  pixel776  pixel777  pixel778  pixel779  pixel780  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0        30        43         0   
3       0  ...         3         0         0         0         0         1   
4       0  ...         0         0         0         0         0         0   

   pixel781  pixel782  pixel783  pixel784  
0         0         0         

(60000, 785)

In [ ]:
#reshaping into 28X28 array
fashion.iloc[3,1:].values.reshape(28,28).astype('uint8')

array([[  0,   0,   0,   1,   2,   0,   0,   0,   0,   0, 114, 183, 112,
         55,  23,  72, 102, 165, 160,  28,   0,   0,   0,   1,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   1,   0,   0,  24, 188, 163,  93, 136, 153,
        168, 252, 174, 136, 166, 130, 123, 131,  66,   0,   0,   1,   0,
          0,   0],
       [  0,   0,   0,   2,   0,  10, 157, 216, 226, 208, 142,  66, 115,
        149, 230, 190, 196, 198, 172, 222, 107, 165, 211,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0, 118, 214, 174, 168, 109, 200, 124, 150,
        143,  58,  63,  89, 137,  97, 168, 138, 142, 195, 156,   0,   0,
          0,   0],
       [  0,   0,   0,   0,  25, 140,  70,  80,  43,  71,  96,  93, 151,
        121, 197, 143, 107,  82, 101, 111,  80, 137, 193, 208,   6,   0,
          0,   0],
       [  0,   0,   0,   0,  74, 194, 107, 146, 178, 185, 182,  77, 185,
        218, 210, 175, 174, 235, 217, 217, 129, 180, 210, 208,  89,   0,
          0,   0],
       [  

<a id = "cell_Preprocessing"></a>
## 2. Data Preprocessing

In [ ]:
#Storing Pixel array in form length width and channel in df_x
df_x = fashion.iloc[:,1:].values.reshape(len(fashion),28,28,1)

#Storing the labels in y
y = fashion.iloc[:,0].values
#Converting labels to categorical features
df_y = keras.utils.np_utils.to_categorical(y,num_classes=10)

df_x = np.array(df_x)
df_y = np.array(df_y)

#lables
y
#categorical labels
df_y
df_x.shape

(60000, 28, 28, 1)

<a id = "cell_model"></a>
## 3. Model Construction

In [ ]:
#test train split
x_train, x_test, y_train, y_test = train_test_split(df_x,df_y,test_size=0.2,random_state=4)

In [ ]:
model = Sequential()
model.add(Convolution2D(32,3,data_format='channels_last',activation='relu',input_shape=(28,28,1)))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Flatten())
model.add(Dense(100))
model.add(Dropout(0.5))
model.add(Dense(10))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer = 'adadelta', metrics = ['accuracy'])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 5408)              0         
                                                                 
 dense (Dense)               (None, 100)               540900    
                                                                 
 dropout (Dropout)           (None, 100)               0         
                                                                 
 dense_1 (Dense)             (None, 10)                1010      
                                                        

<a id = "cell_execution"></a>
## 4. Model Execution

In [ ]:
#fitting it with just 100 images for testing 

model.fit(x_train,y_train, validation_data=(x_test,y_test), epochs=10)

Epoch 1/10
1500/1500 [==============================] - 35s 23ms/step - loss: 5.6739 - accuracy: 0.7026 - val_loss: 2.8912 - val_accuracy: 0.7983
Epoch 2/10
1500/1500 [==============================] - 34s 23ms/step - loss: 5.3875 - accuracy: 0.7102 - val_loss: 2.7845 - val_accuracy: 0.8033
Epoch 3/10
1500/1500 [==============================] - 34s 22ms/step - loss: 5.1455 - accuracy: 0.7205 - val_loss: 2.6960 - val_accuracy: 0.8074
Epoch 4/10
1500/1500 [==============================] - 34s 23ms/step - loss: 4.9280 - accuracy: 0.7271 - val_loss: 2.6301 - val_accuracy: 0.8114
Epoch 5/10
1500/1500 [==============================] - 34s 22ms/step - loss: 4.7559 - accuracy: 0.7337 - val_loss: 2.5428 - val_accuracy: 0.8154
Epoch 6/10
1500/1500 [==============================] - 34s 23ms/step - loss: 4.6329 - accuracy: 0.7365 - val_loss: 2.4809 - val_accuracy: 0.8197
Epoch 7/10
1500/1500 [==============================] - 34s 22ms/step - loss: 4.4885 - accuracy: 0.7415 - val_loss: 2.4502 -

In [ ]:
model.evaluate(x_test,y_test)

375/375 [==============================] - 3s 8ms/step - loss: 2.2881 - accuracy: 0.8281


[2.2881383895874023, 0.828083336353302]

<a id = "cell_report"></a>
## 5. Experiments Report

To evaluate the model we created, it has a accuracy of 82.81% with the predicted data. Therefore, this is fairly a successful prediction model using the convolutional neural network model.